posthog/api/services/query.py

```diff
@@ -13,93 +13,96 @@
 from posthog.models import Team
 from posthog.queries.time_to_see_data.serializers import SessionEventsQuerySerializer, SessionsQuerySerializer
 from posthog.queries.time_to_see_data.sessions import get_session_events, get_sessions
-from posthog.schema import HogQLMetadata
+from posthog.schema import (
+    HogQLMetadata,
+    HogQLQuery,
+    EventsQuery,
+    TrendsQuery,
+    RetentionQuery,
+    QuerySchemaRoot,
+    LifecycleQuery,
+    WebOverviewQuery,
+    WebTopClicksQuery,
+    WebStatsTableQuery,
+    ActorsQuery,
+    SessionsTimelineQuery,
+    DatabaseSchemaQuery,
+    TimeToSeeDataSessionsQuery,
+    TimeToSeeDataQuery,
+    StickinessQuery,
+)
 
 logger = structlog.get_logger(__name__)
 
-QUERY_WITH_RUNNER = [
-    "LifecycleQuery",
-    "RetentionQuery",
-    "TrendsQuery",
-    "StickinessQuery",
-    "WebOverviewQuery",
-    "WebTopSourcesQuery",
-    "WebTopClicksQuery",
-    "WebTopPagesQuery",
-    "WebStatsTableQuery",
-]
-QUERY_WITH_RUNNER_NO_CACHE = [
-    "HogQLQuery",
-    "EventsQuery",
-    "ActorsQuery",
-    "SessionsTimelineQuery",
-]
+QUERY_WITH_RUNNER = (
+    LifecycleQuery
+    | RetentionQuery
+    | StickinessQuery
+    | TrendsQuery
+    | WebOverviewQuery
+    | WebTopClicksQuery
+    | WebStatsTableQuery
+)
+QUERY_WITH_RUNNER_NO_CACHE = HogQLQuery | EventsQuery | ActorsQuery | SessionsTimelineQuery
 
 
-def _unwrap_pydantic(response: Any) -> Dict | List:
-    if isinstance(response, list):
-        return [_unwrap_pydantic(item) for item in response]
-
-    elif isinstance(response, BaseModel):
-        resp1: Dict[str, Any] = {}
-        for key in response.__fields__.keys():
-            resp1[key] = _unwrap_pydantic(getattr(response, key))
-        return resp1
-
-    elif isinstance(response, dict):
-        resp2: Dict[str, Any] = {}
-        for key in response.keys():
-            resp2[key] = _unwrap_pydantic(response.get(key))
-        return resp2
-
-    return response
-
-
-def _unwrap_pydantic_dict(response: Any) -> Dict:
-    return cast(dict, _unwrap_pydantic(response))
+def process_query(
+    team: Team,
+    query_json: dict,
+    limit_context: Optional[LimitContext] = None,
+    refresh_requested: Optional[bool] = False,
+) -> dict:
+    model = QuerySchemaRoot.model_validate(query_json)
+    return process_query_model(
+        team,
+        model.root,
+        limit_context=limit_context,
+        refresh_requested=refresh_requested,
+    )
 
 
-def process_query(
+def process_query_model(
     team: Team,
-    query_json: Dict,
+    query: Any,  # mypy has problems with unions and isinstance
     limit_context: Optional[LimitContext] = None,
     refresh_requested: Optional[bool] = False,
-) -> Dict:
-    # query_json has been parsed by QuerySchemaParser
-    # it _should_ be impossible to end up in here with a "bad" query
-    query_kind = query_json.get("kind")
-    tag_queries(query=query_json)
```

Comment:
@mariusandra not sure what this does, and I don't have a dict anymore that I can pass to it

Reply:
This adds a copy of the input query into ClickHouse's `system.query_log` table.

Reply:
So what can we do? I don't want to serialize again.

Reply:
I tag the queries where I still have the dict.